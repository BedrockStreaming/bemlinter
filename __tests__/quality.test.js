const { lint, format } = require('../src/bemlinter.js');
const fs = require('fs');

const toFilePath = (fileName, type) => `${__dirname}/${type}/${fileName}`;

const snapLintOutput = (lintResult) => {
  const output = format(lintResult, false);
  expect(output).toMatchSnapshot();

  return output;
};

const expectNewError = errorLength => (lintResult) => {
  expect(lintResult.getSnapshot().getNewErrorList().length).toBe(errorLength);

  return lintResult;
};

const expectFileCreated = filePath => (lintResult) => {
  expect(fs.existsSync(filePath)).toBe(true);
  fs.unlinkSync(filePath);

  return lintResult;
};

const getLintResult = (stylesFileName, qualityFileName, options = {}) => {
  options.snapshot = toFilePath(qualityFileName, 'quality');
  return lint(toFilePath(stylesFileName, 'styles'), options);
};

describe('Bemlinter of alright.scss ', () => {
  it('should lint without errors', done => (
    getLintResult('alright.scss', 'alright-snap.json')
      .then(expectNewError(0))
      .then(snapLintOutput)
      .catch(console.error)
      .then(done)
  ));

  it('should create a quality snapshot', done => (
    getLintResult('alright.scss', 'empty-snap.json')
      .then(expectFileCreated(toFilePath('empty-snap.json', 'quality')))
      .catch(console.error)
      .then(done)
  ));
});

describe('Bemlinter of leak.scss ', () => {
  it('should lint without new errors', done => (
    getLintResult('leak.scss', 'leak-snap.json')
      .then(expectNewError(0))
      .then(snapLintOutput)
      .catch(console.error)
      .then(done)
  ));

  it('should lint a new leak error', done => (
    getLintResult('leak.scss', 'alright-snap.json')
      .then(expectNewError(1))
      .then(snapLintOutput)
      .catch(console.error)
      .then(done)
  ));

  it('should create a quality snapshot without new errors', done => (
    getLintResult('leak.scss', 'new-snap.json')
      .then(expectNewError(0))
      .then(expectFileCreated(toFilePath('new-snap.json', 'quality')))
      .catch(console.error)
      .then(done)
  ));
});
