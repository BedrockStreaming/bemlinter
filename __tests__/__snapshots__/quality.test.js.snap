exports[`Bemlinter of alright.scss  should lint without errors 1`] = `
"bemlinter has validated 1 block.
OK: bemlinter has detected no new error."
`;

exports[`Bemlinter of leak.scss should lint two new leak errors 1`] = `
"  ✗ leak
[leak.scss:10] Error: \".external-block\" is incoherent with the file name, \".leak\" expected.
[leak.scss:14] Error: \".another-block\" is incoherent with the file name, \".leak\" expected.

bemlinter has detected 2 errors on 1 block.
FAIL: bemlinter has detected 2 new errors."
`;

exports[`Bemlinter of leak.scss should lint two new leak errors 2`] = `
"  ✗ leak
[leak.scss:10] Error: \".external-block\" is incoherent with the file name, \".leak\" expected.
[leak.scss:14] Error: \".another-block\" is incoherent with the file name, \".leak\" expected.

bemlinter has detected 2 errors on 1 block.
FAIL: bemlinter has detected 2 new errors."
`;

exports[`Bemlinter of leak.scss should lint without new errors 1`] = `
"bemlinter has detected 2 errors on 1 block.
OK: bemlinter has detected no new error."
`;
